Creating and retrieving references

raghav710 opened the pull request:
* Tests for a few common scenarios
* Only Get and Create references implemented as of now as it is required for #84 
* PSScriptAnalyser Run returns no issues

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,219 @@
+        'NoStatus' = (Resolve-ParameterWithDefaultConfigurationValue -Name NoStatus -ConfigValueName DefaultNoStatus)
+    }
+
+    return Invoke-GHRestMethod @params
```
@HowardWolosky this throws a 422 exception if the reference already exists. I was not sure based on other files on what the expected behaviour should be and am hence allowing it to throw the exception (The exception message contains info about why it failed) and am asserting the same in the tests

HowardWolosky replied:
Exception when already exists seems fine to me.  Ok to leave as-is.

raghav710 commented on Tests/GitHubReferences.Tests.ps1:
```diff
@@ -0,0 +1,173 @@
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSAvoidUsingConvertToSecureStringWithPlainText", "", Justification="Needed to configure with the stored, encrypted string value in AppVeyor.")]
+    param()
+
+    if ($env:AppVeyor)
```
I took this portion verbatim from other tests. I would be happy to move this Appveyor bit to a common util file which is then called from other tests. Thoughts?

HowardWolosky replied:
Others have mentioned that idea as well in previous PR's, but no-one ever got around to it.  I think it's a solid idea that I completely agree with, but it should be a different PR and not part of this one.  By all means, you're welcome to make that change as well, but please keep that as a separate PR.

HowardWolosky replied:
I did actually do this in a later change, so there will be some merging/resolving that you'll need to do as part of this PR.

raghav710 commented:
I'm planning to rebase my #84 branch on top of this so I can get that ready for review as well while this is being reviewed

HowardWolosky commented on PowerShellForGitHub.psd1:
```diff
@@ -115,7 +116,9 @@
         'Update-GitHubLabel',
-        'Update-GitHubRepository'
+        'Update-GitHubRepository',
+        'Get-GitHubReference',
```
Please keep these alphabetical.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,227 @@
+        Write-Log -Message $message -Level Error
+        throw $message
+    }
+    $uriFragment = "/repos/$OwnerName/$RepositoryName/git/refs/$Reference"
```
nit: line-feed after a closing statement (`}`)

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,227 @@
+        [string] $Uri,
+
+        [Parameter(Mandatory)]
+        [string] $Reference,
```
If you make this optional (leaving off `[Parameter(Mandatory)]`, then this will handle _both_ a [single reference](https://developer.github.com/v3/git/refs/#get-a-reference) and [all references](https://developer.github.com/v3/git/refs/#get-all-references), but you'd also have to call `Invoke-GHRestMethodMultipleResult` and branch on what you set `$description` to.  This is a common pattern you can see in other methods (like `Get-GitHubIssue`)

raghav710 replied:
Thanks for the note, I'll have a look at the other methods and adapt the same approach here

raghav710 replied:
@HowardWolosky looks like the API/docs have been updated and the single/all reference links are no longer available. Can I understand getting a reference to mean _getting a single reference_ and getting all references to mean _listing matching references_? In that case should I create a new method for listing matching references?

HowardWolosky replied:
Looks like it would now be [List Matching References](https://developer.github.com/v3/git/refs/#list-matching-references).

HowardWolosky replied:
Some thought needs to go in to how you would make a combined single/multiple function given the additional suggestion of splitting out `$Reference` as discussed below.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,227 @@
+
+    $telemetryProperties = @{
+        'OwnerName' = (Get-PiiSafeString -PlainText $OwnerName)
+        'RepositoryName' = (Get-PiiSafeString -PlainText $RepositoryName)
```
Assuming you make the change to have this support _both_ individual and _multiple_ lookup, it would be helpful here to have a boolean property to track if the user provided a `Reference`, like this:

https://github.com/PowerShell/PowerShellForGitHub/blob/68fa5a563b96490774e9c5fae38193194f765157/GitHubIssues.ps1#L190

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,227 @@
+    Details of the git reference created. Throws an Exception if the reference already exists
+
+    .EXAMPLE
+        -GitHubReference -OwnerName Powershell -RepositoryName PowerShellForGitHub -Reference refs/heads/master -Sha aa218f56b14c9653891f9e74264a383fa43fefbd
```
`-GitHubReference` -- typo (missing `New`)

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,227 @@
+        If not supplied here, the DefaultNoStatus configuration property value will be used.
+
+    .OUTPUTS
+    [PSCustomObject]
```
nit: indent

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,227 @@
+    $description =  "Creating Reference $Reference for $RepositoryName from SHA $Sha"
+
+    $hashBody = @{
+        'ref' = $Reference
```
this should prepend `refs/`, users shouldn't have to deal with that part of the reference.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,227 @@
+        'Method' = 'Post'
+        'Body' = (ConvertTo-Json -InputObject $hashBody)
+        'Description' =  $description
+        'AcceptHeader' = 'application/vnd.github.symmetra-preview+json'
```
This preview header isn't valid for this API.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,227 @@
+        'UriFragment' = $uriFragment
+        'Method' = 'Get'
+        'Description' =  $description
+        'AcceptHeader' = 'application/vnd.github.symmetra-preview+json'
```
This preview header isn't valid for this API.

HowardWolosky commented on Tests/GitHubReferences.Tests.ps1:
```diff
@@ -0,0 +1,171 @@
+        Describe 'Create a new reference(branch) in repository' {
+            $repositoryName = [Guid]::NewGuid()
+            $repo = New-GitHubRepository -RepositoryName $repositoryName -AutoInit
+            $existingref = @(Get-GitHubReference -OwnerName $ownerName -RepositoryName $repositoryName -Reference "heads/master")
```
Why are you wrapping the result in array syntax? `@()`
Same question for here and elsewhere in this file.

raghav710 replied:
I probably saw this elsewhere and adopted it here. removed it now

HowardWolosky commented on Tests/GitHubReferences.Tests.ps1:
```diff
@@ -0,0 +1,171 @@
+            $repo = New-GitHubRepository -RepositoryName $repositoryName -AutoInit
+
+            Context 'On getting a valid reference from a new repository' {
+                $reference = @(Get-GitHubReference -OwnerName $ownerName -RepositoryName $repositoryName -Reference "heads/master")
```
it would probably be worthwhile to have a constant for `head/master` that you use here and elsewhere, and then when testing against it, you could just prepend `refs/` to that same value.

HowardWolosky commented on Tests/GitHubReferences.Tests.ps1:
```diff
@@ -0,0 +1,171 @@
+                    $reference | Should be $null
+                }
+            }
+            $null = Remove-GitHubRepository -OwnerName $ownerName -RepositoryName $repositoryName
```
nit: line-feed after a closing statement (`}`)

raghav710 commented:
@HowardWolosky dropping a note to update that I'll take a look at this probably this weekend. (I wish I could spend more time on this repo)
Thanks for being patient!

HowardWolosky commented:
Thanks for the update. No worries.  I'll keep this PR open as long as is needed.  I appreciate the contribution.

raghav710 commented:
@HowardWolosky  no need to review this yet, will make a few more changes and address remaining comments before its ready for review again. Thanks for being patient, happy to be back!

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,241 @@
+        SupportsShouldProcess,
+        DefaultParametersetName='Elements')]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSShouldProcess", "",
+    Justification="Methods called within here make use of PSShouldProcess, and the switch is passed on to them inherently.")]
```
Moved this to the next line because this was too far away from the 100 column limit. I haven't seen this being done in other files though. Should I keep it the old way to make it consistent? Thoughts?

raghav710 commented on Tests/GitHubReferences.Tests.ps1:
```diff
@@ -0,0 +1,200 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
This file has been refactored a bit, mostly with respect to usage of names like "heads/master" which have been replaced with variables

raghav710 commented on Tests/GitHubReferences.Tests.ps1:
```diff
@@ -0,0 +1,201 @@
+                $reference = Get-GitHubReference -OwnerName $ownerName -RepositoryName $repositoryName
+
+                It 'Should return all references' {
+                    ($reference.ref | Where-Object {$refNames -Contains $_}).Count | Should be $refNames.Count
```
This felt like a little too much programming for a test case, but I wasn't happy with checking for the count alone either. Should I just manually check if both items are present or is this fine?

HowardWolosky replied:
I think this is fine.  Thanks for checking.

raghav710 commented:
@HowardWolosky this is now ready for another round of review. I havent added the Update and delete bits yet and plan to do so once I'm done with my other Pull request, so this is kind of minimal

HowardWolosky commented on Tests/GitHubReferences.Tests.ps1:
```diff
@@ -0,0 +1,201 @@
+            $repositoryName = [Guid]::NewGuid()
+            $repo = New-GitHubRepository -RepositoryName $repositoryName -AutoInit
+            $masterRefName = "heads/master"
+            $randomRefName = "heads/someRandomRef"
```
```suggestion
            $randomRefName = "heads/$([Guid]::NewGuid())"
```

This way the tests are truly robust.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,242 @@
+            Write-Log -Level Error "No reference named $Reference exists in repository $RepositoryName"
+            return $null
+        }
+        Write-InteractiveHost "No references exist in repository $RepositoryName" -NoNewline -f Red
```
In general, I think I want to avoid having individual functions using `Write-InteractiveHost`.  Writing it to `Error` via `Write-Log` should be enough to communicate to users, then `throw`.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,242 @@
+    }
+    catch
+    {
+        if ($PSBoundParameters.ContainsKey('Reference'))
```
I don't think you should be doing this additional error checking at all.  If you just let the natural error flow occur (e.g. just `return Invoke-GHRestMethodMultipleResult @params`) and don't bother putting it within this `try/catch`.

raghav710 commented:
@HowardWolosky thanks for the comments. I'll make the required changes and submit for review

raghav710 commented:
@HowardWolosky I've incorporated your comments and added the update and delete methods. This is ready for another round of review
However I wasn't able to test the update method's positive scenario and would appreciate any help on the same.

I tried creating a couple of branches in a sample repo (https://github.com/raghav710/61603191-d09d-4a1a-8a52-88e719371d8c/pulls) and tried to update the SHA of one of the branches to that of the other. But that didnt work. 

Also, I rebased the changes on top of master and later pulled from this branch which caused some issues (started showing 26 files as changed), hence I squashed the commits and after ensuring it is rebased on top of latest master, have force pushed it. Sorry if this makes the review difficult

raghav710 commented:
@HowardWolosky  these tests require the Contents API to be implemented. Have created an issue for that and will be working on it

raghav710 commented:
@HowardWolosky this is now ready for review. Please let me know your comments on this.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+            ParameterSetName='Uri')]
+        [string] $Uri,
+
+        [string] $Reference,
```
As I've thought about this more, I don't think it makes sense for users to have to know the syntax `heads/<branch name>` or `tags/<tag name>`.  Instead, I think you should split this out and have an optional parameter of `[string] $Tag` and another one as `[string] $Head`.  Users can just specify the one that should be used. You can defined them to be mutually exclusive via parametersets.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+
+    if ($OwnerName -xor $RepositoryName)
+    {
+        $message = 'You must specify both Owner Name and Repository Name.'
```
That's not what `-xor` does.  Your condition, as written right now means:
"Either OwnerName or RepositoryName must have a value, but BOTH can't have a value."
So, either your condition is wrong or your error message is wrong.

That being said, you don't need to verify this at all.  By default, `Resolve-RepositoryElements` will throw an exception if it isn't returning with a non-empty OwnerName and RepositoryName. Linking to the documentation: https://github.com/microsoft/PowerShellForGitHub/blob/6735ba57a5a43b61a37ef09d4021296dcd417dba/GitHubCore.ps1#L784-L788

raghav710 replied:
I picked up this check from [GitHubIssues.ps1](https://github.com/microsoft/PowerShellForGitHub/blob/master/GitHubIssues.ps1#L195) where there is a validation check to ensure that both OwnerName and RepositoryName are provided and would throw an error message (as is done here) when only one of them is provided (hence the xor)

I agree on doing away with this validation as Resolve-RepositoryElements takes care of it anyway!

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+        throw $message
+    }
+
+    if (-not [String]::IsNullOrEmpty($RepositoryName))
```
We'll never be in a situation where we don't have a repo name, per the above.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+
+    if (-not [String]::IsNullOrEmpty($RepositoryName))
+    {
+        $uriFragment = "repos/$OwnerName/$RepositoryName/git/matching-refs"
```
Where are you handling the [singular reference](https://developer.github.com/v3/git/refs/#get-a-single-reference) case?

raghav710 replied:
Updated this in the latest commit

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+        else
+        {
+            # Add a slash at the end. Calling it with only matching-refs without the slash causes a 404
+            $uriFragment = $uriFragment + "//"
```
If this is true, then the logic above is a bit delicate.  What if `$Reference` has a leading slash?  What happens then?  I think you need to do some more robust input validation.

raghav710 replied:
The additional slash at the end is required because Invoke-GHRestMethod removes trailing slashes.
https://github.com/microsoft/PowerShellForGitHub/blob/d279f1c7e1bd8261cdcf9dab6a8b78d3a0004457/GitHubCore.ps1#L126-L130
In this case, since there would not be any other variable appended, no input validation is required
And for the part where we do
`$uriFragment = $uriFragment + "/$Reference"`

Validation would not be required since the $reference variable would now be generated from the Tag/Branch passed in by the user

Please feel free to correct me if I misunderstood this comment

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+        [string] $Uri,
+
+        [Parameter(Mandatory)]
+        [string] $Reference,
```
Same feedback as for `Get-*`.  Instead of making users understand the `head/<branch name>` / `tag/<tag name>` syntax, why not just provide explicit parameters that can be used?

raghav710 replied:
This is now done

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+        'RepositoryName' = (Get-PiiSafeString -PlainText $RepositoryName)
+    }
+
+    if ($OwnerName -xor $RepositoryName)
```
Same feedback as above.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+        'RepositoryName' = (Get-PiiSafeString -PlainText $RepositoryName)
+    }
+
+    if ($OwnerName -xor $RepositoryName)
```
Same feedback as above.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+        [string] $Uri,
+
+        [Parameter(Mandatory)]
+        [string] $Reference,
```
Same feedback as above.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+        [string] $Uri,
+
+        [Parameter(Mandatory)]
+        [string] $Reference,
```
Same feedback as above.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+        'RepositoryName' = (Get-PiiSafeString -PlainText $RepositoryName)
+    }
+
+    if ($OwnerName -xor $RepositoryName)
```
Same feedback as above.

HowardWolosky commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,442 @@
+    $uriFragment = "repos/$OwnerName/$RepositoryName/git/refs/$Reference"
+    $description =  "Deleting Reference $Reference from repository $RepositoryName"
+
+    $params = @{
```
A recent PR introduced the idea of getting user confirmation on delete actions (which can be overridden by a commandline parameter).  I think it's a pretty solid idea, and I think we should apply that approach on all new "delete" methods.   Take a [look here](https://github.com/microsoft/PowerShellForGitHub/commit/536b762425d51a181166c2c47ad2b00014911d1d#diff-2a0b254e60269a3a0794d63c7922522dR518-R549) for more info.

I'd ask that you _also_ add an additional Example to here that shows how to avoid the prompt (as that reference that I pointed to does as well).

raghav710 replied:
Agreed. That looks like a good idea. 
suggestion: Can we have the flag as -NoConfirm or similar so that the user need not type in -Confirm:$false ?

raghav710 commented:
Thanks for the comments @HowardWolosky  I'll take a look at the changes requested and try to get back as soon as possible. Sorry it has taken a year for this since I wasn't continuously involved

HowardWolosky commented:
@raghav710 -- can you confirm that this is ready for re-review?

Additionally, I used GitHub's merge conflict resolver for you on the manifest file, and it appears that it messed up the line breaks in the file, causing the whole file to appear to have been changed.  Can you please push two new commits, one where you revert the file to its original state, and the second where you re-apply your specific changes, so that only the changes are seen in the diff?  Thanks!

raghav710 commented:
@HowardWolosky sure. I'll do that. I checked the tests and i remember them running fine. Will check the code once again for any mistakes and planning to drop a note max by tomorrow that this is ready for review. Sorry this has taken so long

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+    .PARAMETER All
+        If provided, will retrieve both branches and tags for the given repository
+
+    .PARAMETER MatchPrefix
```
This is the option for calling the matching-refs API. Wasn't sure what is the standard pattern for such a functionality

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+        Get the branch 'powershell' and if it doesn't exist, get all branches beginning with 'powershell'
+
+    .EXAMPLE
+        Get-GitHubReference -Uri https://github.com/You/YourRepo -All
```
I had to add an 'All' parameter because of the parametersets. Since both branch and tag are optional, calling Get-GithubReference without parameters results in PS not being able to find a parameterset, hence this workaround. Appreciate any suggestions on bettering this

raghav710 replied:
Thanks. I think this greatly simplifies it!

raghav710 commented:
@HowardWolosky this is now ready for another round of review. Sorry there are multiple commits, I would like to squash these before merging.

Please let me know your thoughts on the Get-GithubReference, I had to take a few decisions with the parameters to accomodate the different use cases. Have also updated the tests to be more extensive.

raghav710 commented:
@HowardWolosky please feel free to comment on the approach used here and whether it is good enough. Also if you think some of this functionality is already being added by #200 I'm fine with removing parts of these changes so we can bring in only what is required

raghav710 commented:
Thanks for the comments @HowardWolosky . I would take some time in getting back to this. Is there a timeline we are looking at to get this merged? (So the dependent PRs dont have to wait?) and I can try working with that in mind

HowardWolosky commented:
> Thanks for the comments @HowardWolosky Howard Wolosky FTE . I would take some time in getting back to this. Is there a timeline we are looking at to get this merged? (So the dependent PRs dont have to wait?) and I can try working with that in mind

Do you think you can get to it within the next two weeks?  If not, then maybe someone else in the community might be able to apply the feedback to what you've done so far, in order to unblock the other PR's waiting on this (#200 is definitely waiting on this, but there are others like #193 which need it for their UT's).

You've put a bunch of effort into this so far, and I want to ensure that you're getting the appropriate credit for your work.  I also want to keep the momentum going for the others that are also actively contributing to the project and prevent blockages from happening for too long.

Thanks for your understanding and your participation!

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+
+}
+
+function Resolve-Reference($Tag, $Branch)
```
@HowardWolosky does this apply to helper functions which we'd not want to expose directly to the user as well. Is there a different convention I should use for such functions?

HowardWolosky replied:
It applies to _all_ functions, even those not exported.  The CBH can be just as helpful for the maintainer as it can be for the end-user.

raghav710 replied:
Agreed. Will update this one

raghav710 commented:
@HowardWolosky I plan to work on this today and fix the code based on comments

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+        DefaultParameterSetName='Elements')]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSShouldProcess", "", Justification="Methods called within here make use of PSShouldProcess, and the switch is passed on to them inherently.")]
+    param(
+        [Parameter(Mandatory, ParameterSetName='BranchElements')]
```
Will update this

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+        'RepositoryName' = (Get-PiiSafeString -PlainText $RepositoryName)
+    }
+
+    $reference = Resolve-Reference $Tag $Branch
```
Will fix this as well as ensuring there is CBH for the method

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+    return Invoke-GHRestMethod @params
+}
+
+function Update-GitHubReference
```
Fixed in the next commit

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+
+        Remove the tag milestone1 without prompting for confirmation
+    #>
+    [CmdletBinding(SupportsShouldProcess, ConfirmImpact="High")]
```
Fixed in the next commit

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+
+    $repositoryInfoForDisplayMessage = if ($PSCmdlet.ParameterSetName -eq "Uri") { $Uri } else { $OwnerName, $RepositoryName -join "/" }
+    $reference = Resolve-Reference $Tag $Branch
+    if ($PSCmdlet.ShouldProcess($repositoryInfoForDisplayMessage, "Remove reference: $reference"))
```
Fixed in next commit

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+
+        return Invoke-GHRestMethod @params
+    }
+
```
Fixed in next commit

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+
+    Write-InvocationLog -Invocation $MyInvocation
+
+    $elements = Resolve-RepositoryElements -BoundParameters $PSBoundParameters -DisableValidation
```
Thanks for the note. As mentioned in #243 this is one of the copy-paste issues :) will fix this

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+    $telemetryProperties = @{
+        'OwnerName' = (Get-PiiSafeString -PlainText $OwnerName)
+        'RepositoryName' = (Get-PiiSafeString -PlainText $RepositoryName)
+        'ProvidedReference' = $PSBoundParameters.ContainsKey('Reference')
```
Removed this. was an artifact of the previous code where branches/tags were not parameters

raghav710 commented on GitHubReferences.ps1:
```diff
@@ -0,0 +1,525 @@
+    $uriFragment = "repos/$OwnerName/$RepositoryName/git"
+
+    if ($All) {
+        # Add a slash at the end as Invoke-GHRestMethod removes the last trailing slash. Calling this API without the slash causes a 404
```
I am open for any suggestions to making this better. At this point, this is all I'm able to come up with :(

raghav710 commented on Tests/GitHubReferences.Tests.ps1:
```diff
@@ -0,0 +1,466 @@
+$moduleRootPath = Split-Path -Path $PSScriptRoot -Parent
+. (Join-Path -Path $moduleRootPath -ChildPath 'Tests\Common.ps1')
+
+function Update-MasterSHA()
```
Updated to use Set-Content

raghav710 commented:
@HowardWolosky I have made an attempt at adding pipeline support and refactored where required to match the current standards. Also updated the tests to check for pipelines and to follow Pester standards where applicable.

Taking this forward, would someone in the community be willing to take this from here? Given that there are other PRs waiting on this, I dont want to delay this any longer waiting for my availability.

PS: its amazing how much has been done in the project in a short period! from pipeline support to removing boilerplate from tests to standardizing many code conventions. Kudos to you and the community!

HowardWolosky commented:
Thanks for all your efforts here, @raghav710.  You've provided a solid base for the remaining work that needs to be done.  I'm nearing completion on an update to this work to fill in the remaining gaps.  Once I'm done with that, I'll close out this PR and start a new one.  Again, thanks so much.  Looking forward to seeing your further contributions in this project when you have time.

raghav710 commented:
Thanks @HowardWolosky for giving me a chance to contribute to this project. Again, apologies for not taking this to completion
